### PR TITLE
[FLINK-8944] [Kinesis Connector] Use listShards instead of DescribeSt…

### PR DESCRIPTION
**Pull request by kailashhd**
…ream for shard discovery as it offer higher rate limits

## What is the purpose of the change

List Shards provides high AWS rate limits unlike DescribeStreams (which is on AWS account level) allowing faster shard discovery when kinesis data source in case streams are changed(re-sharded)

## Brief change log
 - Change the kinesis connector to use listShards instead of DescribeStream for shard discovery.

## Verifying this change

This change added tests and can be verified as follows:
 - Added a unit test to check the code path mocking out the kinesis depenedencies
 - Tested by running a small flink job with this connector.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/config/ConsumerConfigConstants.java [hunk 1]
```diff
@@ -66,13 +66,13 @@ public SentinelSequenceNumber toSentinelSequenceNumber() {
 	public static final String STREAM_TIMESTAMP_DATE_FORMAT = "flink.stream.initpos.timestamp.format";
 
 	/** The base backoff time between each describeStream attempt. */
-	public static final String STREAM_DESCRIBE_BACKOFF_BASE = "flink.stream.describe.backoff.base";
```
Comment (tweise) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/config/ConsumerConfigConstants.java]: This would be a breaking change. We should leave these properties as is, if they are semantically equivalent.

Comment (kailashhd) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/config/ConsumerConfigConstants.java]: I have retained the properties string values as same but changed the variable names to indicate listShards for readability.

**Comment by kailashhd** [comment]
R: @tzulitai @tweise

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/config/ConsumerConfigConstants.java [hunk 2]
```diff
@@ -65,14 +65,14 @@ public SentinelSequenceNumber toSentinelSequenceNumber() {
 	/** The date format of initial timestamp to start reading Kinesis stream from (when AT_TIMESTAMP is set for STREAM_INITIAL_POSITION). */
 	public static final String STREAM_TIMESTAMP_DATE_FORMAT = "flink.stream.initpos.timestamp.format";
 
-	/** The base backoff time between each describeStream attempt. */
-	public static final String STREAM_DESCRIBE_BACKOFF_BASE = "flink.stream.describe.backoff.base";
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/config/ConsumerConfigConstants.java]: Changing the name of this variable, strictly speaking, breaks backwards compatibility, as users might be using them.

Comment (tzulitai) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/config/ConsumerConfigConstants.java]: This goes the same for all other key variable rename changes in this class.
I would suggest to deprecate existing ones if we want to change the name internally.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java [hunk 3]
```diff
@@ -382,50 +386,62 @@ protected static boolean isRecoverableException(AmazonServiceException ex) {
...
+				LOG.warn("List Shards has an expired token. Reusing the previous state.");
+				break;
 			}
 		}
-
-		// Kinesalite (mock implementation of Kinesis) does not correctly exclude shards before the exclusive
-		// start shard id in the returned shards list; check if we need to remove these erroneously returned shards
-		if (startShardId != null) {
-			List<Shard> shards = describeStreamResult.getStreamDescription().getShards();
+    // Kinesalite (mock implementation of Kinesis) does not correctly exclude shards before
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java]: incorrect indentation

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java [hunk 4]
```diff
@@ -382,50 +386,62 @@ protected static boolean isRecoverableException(AmazonServiceException ex) {
...
+		// the exclusive start shard id in the returned shards list; check if we need to remove
+		// these erroneously returned shards.
+		if (startShardId != null && listShardsResults != null) {
+			List<Shard> shards = listShardsResults.getShards();
 			Iterator<Shard> shardItr = shards.iterator();
-			while (shardItr.hasNext()) {
+			while (shardItr.hasNext()){
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java]: code style nit: missing space before `{`

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java [hunk 5]
```diff
@@ -353,19 +355,21 @@ protected static boolean isRecoverableException(AmazonServiceException ex) {
...
-		DescribeStreamResult describeStreamResult;
+		// List Shards returns just the first 1000 shard entries. In order to read the entire stream,
+		// we need to use the returned nextToken to get additional shards.
+		ListShardsResult listShardsResult;
+		String startShardToken = null;
 		do {
-			describeStreamResult = describeStream(streamName, lastSeenShardId);
-
-			List<Shard> shards = describeStreamResult.getStreamDescription().getShards();
+			listShardsResult = listShards(streamName, lastSeenShardId, startShardToken);
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java]: if within `listShards(...)` we caught the `ExpiredNextTokenException`, then `null` will be returned as the result, correct?
If so, then the current built up `shardsIfStream` will be returned immediately, regardless of whether or not there are more shards following.

Although it might not be too common that we have expired tokens here, I wonder if we can handle this case more gracefully (e.g., re-fetching a token to make sure that there really is no more shards).

File: flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyTest.java [hunk 6]
```diff
@@ -70,20 +113,107 @@ public void testIsRecoverableExceptionWithNullErrorType() {
...
+
+		ListShardsRequestMatcher(String shardIdArg, String nextTokenArg) {
+			shardId = shardIdArg;
+			nextToken = nextTokenArg;
+		}
+
+		@Override
+		protected boolean matchesSafely(final ListShardsRequest listShardsRequest, final Description description) {
+			if (shardId == null) {
+				if (StringUtils.isNotEmpty(listShardsRequest.getExclusiveStartShardId())) {
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyTest.java]: Can we avoid using `StringUtils` and just use `!String.isEmpty()` instead?

**Comment by tweise** [comment]
Looks like we can either leave all the application facing constants as they are (which might look a bit strange given they refer to old API and we are now using the new), or we deprecate and duplicate :) If the latter, then I would suggest to map from old/deprecated to new property keys in a single place and issue deprecation warning.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java (hunk 3 above)
Comment (kailashhd) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java]: Done.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java (hunk 4 above)
Comment (kailashhd) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java]: Done.

File: flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyTest.java (hunk 6 above)
Comment (kailashhd) [reply flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyTest.java]: Done. Also note that there is quite a bit of code from reused from here: https://goo.gl/MGSRb3. Do I need to add reference to that code?

File: flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyTest.java [hunk 7]
```diff
@@ -97,8 +228,7 @@ public void testClientConfigOverride() {
 
 		AmazonKinesis kinesisClient = Whitebox.getInternalState(proxy, "kinesisClient");
 		ClientConfiguration clientConfiguration = Whitebox.getInternalState(kinesisClient,
-			"clientConfiguration");
+						"clientConfiguration");
```
Comment (tweise) [inline flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyTest.java]: remove unnecessary change

File: flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyTest.java [hunk 8]
```diff
@@ -97,8 +228,7 @@ public void testClientConfigOverride() {
...
 		assertEquals(9999, clientConfiguration.getSocketTimeout());
 	}
-
```
Comment (tweise) [inline flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyTest.java]: remove unnecessary change

File: flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyTest.java (hunk 7 above)
Comment (kailashhd) [reply flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyTest.java]: Done. Sorry for the inconvenience.

File: flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyTest.java (hunk 8 above)
Comment (kailashhd) [reply flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyTest.java]: Done. Sorry for the inconvenience.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/config/ConsumerConfigConstants.java (hunk 2 above)
Comment (kailashhd) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/config/ConsumerConfigConstants.java]: I have deprecated some old properties and added the new ones. This is my first time using the @deprecated annotations. Do let me know if there are better ways of doing this. I used sample PRs like https://goo.gl/LWcrp2 for these changes.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/util/KinesisConfigUtil.java [hunk 9]
```diff
@@ -181,6 +181,25 @@ public static Properties replaceDeprecatedProducerKeys(Properties configProps) {
 		return configProps;
 	}
 
+	public static Properties replaceDeprecatedConsumerKeys(Properties configProps) {
+		if (configProps.containsKey(ConsumerConfigConstants.STREAM_DESCRIBE_BACKOFF_BASE)) {
```
Comment (tweise) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/util/KinesisConfigUtil.java]: Minor: this could be generalized by iterating over a map of oldkey -> newkey and I would also suggest to log a warning for deprecated keys

File: flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyTest.java [hunk 10]
```diff
@@ -26,20 +29,60 @@
...
+import static org.mockito.Mockito.mock;
 
 /**
  * Test for methods in the {@link KinesisProxy} class.
  */
 public class KinesisProxyTest {
+	private static final String NEXT_TOKEN = "NextToken";
+	private static final String fakeStreamName = "fake-stream";
+	private Set<String> shardIdSet;
+	private List<Shard> shards;
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyTest.java]: Should we move these to be scoped only to the `testGetShardList ` test method?

File: flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyTest.java [hunk 11]
```diff
@@ -26,20 +29,60 @@
...
+			initial.put(stream, null);
+		}
+		return initial;
+	}
+
+	private static ListShardsRequestMatcher initialListShardsRequestMatcher() {
+		return new ListShardsRequestMatcher(null, null);
+	}
+
+	private static ListShardsRequestMatcher listShardsNextToken(final String nextToken) {
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyTest.java]: nit: IMO, it would help with readability if we move these private utility methods after the main test ones.

Comment (kailashhd) [reply flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyTest.java]: Done

File: flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyTest.java (hunk 10 above)
Comment (kailashhd) [reply flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyTest.java]: Done

**Comment by kailashhd** [comment]
Please note I rebased my changes to the master as I was having some failures when building just the flink-connector-kinesis maven project. I hope this will not cause problems with reviewing the code changes. Sorry for the inconvenience if it does. :(

**Comment by kailashhd** [comment]
Bumping this up in case it got lost in the queue. @tzulitai
